### ISSUE TITLE
spoken_to matcher

### ISSUE THREAD

**Issue by go8ose** [opsdroid/opsdroid#273]
I've been thinking about the context issue, #156.  I think another element of context is whether or not the bot is directly addressed by the user.  You might only want a skill to be invoked if a user messages the bot in a direct one-to-one channel, or specifically mentions the bots name in the message.

For example, you might want to distinguish between these two interactions:
john> @bot can you please Delete all the things?

…
john> and then, can you believe it, this guy said to his home assistant "Delete all the things" and that silly bot went ahead and did it!

An example matcher might be:

```python
@match_spoken_to()
@match_regex("Delete all the things")
async def delete(opsdroid, config, message):
    opsdroid.context.update_session('question', 'delete-everything')
    await message.respond('Are you sure?')
…
```

This could be part of a new match_context(), but it feels fairly specific. I wonder if the there might be a family of matchers under the 'context' umbrella, instead of one matcher.

**Comment by jacobtomlinson** [comment]
Yeah I've thought about this before, but this seems like a neat approach!

I know in other bot frameworks like Hubot there is a distinction between `hear` and `respond` where the first is for overhearing within general chat and the second is direct messages only.

Just thinking about how this could be implemented. The `Message` object could have a `direct` parameter which is boolean and is set by the connector. This way the connector can parse whether the message was said in a private message or whether the message was prefixed by the bot's username. This would be a breaking change to the `Message` and `Connector` classes.

**Comment by jacobtomlinson** [comment]
Also not sure if I like `@match_spoken_to`, perhaps there is a better way to phrase it.

**Comment by jacobtomlinson** [comment]
To build on #933 we could add a parse step which infers whether a message was intended to be direct to the bot. 

This could be by having the bots name at the beginning of the message, being in a DM room with the bot or maybe even by some more advanced NLU. 

This parser would add a property to the event that it was direct which could then be filtered on with a constraint.
